server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -0,0 +1,170 @@
+        Set<ShardRouting> inEligibleShards,
+        RoutingAllocation allocation
+    );
+
+    protected FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> fetchData(
+        ShardRouting shard,
+        RoutingAllocation allocation
+    ) {
+        logger.error("fetchData for single shard called via batch allocator");
+        throw new IllegalStateException("PrimaryShardBatchAllocator should only be used for a batch of shards");
+    }
+
+    @Override
+    public AllocateUnassignedDecision makeAllocationDecision(ShardRouting unassignedShard, RoutingAllocation allocation, Logger logger) {
+
+        return makeAllocationDecision(new HashSet<>(Collections.singletonList(unassignedShard)), allocation, logger).get(unassignedShard);
+    }
+
+    /**
+     * Build allocation decisions for all the shards given to this allocator.
+     *
+     * @param shards     set of shards given for allocation
+     * @param allocation current allocation of all the shards
+     * @param logger     logger used for logging
+     * @return shard to allocation decision map
+     */
+    @Override
+    public HashMap<ShardRouting, AllocateUnassignedDecision> makeAllocationDecision(
+        Set<ShardRouting> shards,
+        RoutingAllocation allocation,
+        Logger logger
+    ) {
+        HashMap<ShardRouting, AllocateUnassignedDecision> shardAllocationDecisions = new HashMap<>();
+        final boolean explain = allocation.debugDecision();
+        Set<ShardRouting> eligibleShards = new HashSet<>();
+        Set<ShardRouting> inEligibleShards = new HashSet<>();
+        // identify ineligible shards
+        for (ShardRouting shard : shards) {
+            AllocateUnassignedDecision decision = getInEligibleShardDecision(shard, allocation);
+            if (decision != null) {
+                inEligibleShards.add(shard);
+                shardAllocationDecisions.put(shard, decision);
+            } else {
+                eligibleShards.add(shard);
+            }
+        }
+        // Do not call fetchData if there are no eligible shards
+        if (eligibleShards.isEmpty()) {
+            return shardAllocationDecisions;
+        }
+        // only fetch data for eligible shards
+        final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(
+            eligibleShards,
+            inEligibleShards,
+            allocation
+        );
+        // Note : shardsState contain the Data, there key is DiscoveryNode but value is Map<ShardId,
+        // NodeGatewayStartedShardsBatch> so to get one shard level data (from all the nodes), we'll traverse the map
+        // and construct the nodeShardState along the way before making any allocation decision. As metadata for a
+        // particular shard is needed from all the discovery nodes.
+
+        // process the received data
+        for (ShardRouting unassignedShard : eligibleShards) {
+            if (shardsState.hasData() == false) {
```

Comment:
Can we get rid of this code and use the new `NodeGatewayStartedShards` and form a map with old `NodeGatewayStartedShards`. Since we are already creating new map/list of `NodeShardState` this shouldn't change anything in terms of complexity and it will help in avoiding redundant checks in child class.

Reply:
Using the `NodeGatewayStartedShards` class while giving data to PSA.